### PR TITLE
mammon.ext.ircv3.sasl: Allow 400-length continuation lines

### LINKED ISSUE #87: Accept SASL AUTHENTICATE continuation lines
SASL allows clients to send really long `AUTHENTICATE` strings to servers by sending multiple lines.

Right now we simply process the first line and fail the authentication. We should be accepting and processing these longer lines.

I guess we should have a limit on how many consecutive 400-character `AUTHENTICATE` messages we allow being sent to us before we fail the auth outright, 3-5 maybe? I'm not sure that's specified anywhere, or how other implementations handle that.

See also #70

### PR DESCRIPTION
**Pull request by DanielOaks**
This lets the `AUTHENTICATE` command accept the 400-length continuation lines. This allows four 400-length lines and one more line (either just a `+` or the rest of the password).

If they send more 400-length lines after those first four, it will reject it for length. We can change that without too much trouble, but it's probably good to limit it to stop it from just endlessly growing.

Fixes #87 and #70.

Also threw in a small fix in `m_sasl_plain` to dump `904` when they send an invalid line.

### CODE REVIEW

File: mammon/ext/ircv3/sasl.py [hunk 1]
```diff
@@ -40,23 +40,45 @@ def m_AUTHENTICATE(cli, ev_msg):
...
             cli.dump_numeric('905', ['SASL message too long'])
             cli.sasl = None
+            cli.sasl_tmp = ''
+            return
+        elif len(raw_data) == 400:
+            if not hasattr(cli, 'sasl_tmp'):
+                cli.sasl_tmp = ''
+            cli.sasl_tmp += raw_data
+            # allow 4 'continuation' lines before rejecting for length
+            if len(cli.sasl_tmp) > 400 * 4:
```
Comment (progval) [inline mammon/ext/ircv3/sasl.py:62]: Could you make this configurable?

File: mammon/ext/ircv3/sasl.py [hunk 2]
```diff
@@ -40,23 +40,45 @@ def m_AUTHENTICATE(cli, ev_msg):
...
+                cli.dump_numeric('904', ['SASL authentication failed: Password too long'])
+                cli.sasl = None
+                cli.sasl_tmp = ''
             return
+        elif getattr(cli, 'sasl_tmp', None):
+            if raw_data != '+':
+                cli.sasl_tmp += raw_data
```
Comment (progval) [inline mammon/ext/ircv3/sasl.py:69]: Wouldn't this crash if `irc.sasl_tmp` is unset?

Comment (DanielOaks) [inline mammon/ext/ircv3/sasl.py:69]: That getattr just above makes sure it's set (and whenever we 'unset' it, we
set it to an empty string anyways)

Comment (progval) [inline mammon/ext/ircv3/sasl.py:69]: Oh, indeed, I misread it.

**Comment by progval** [comment]
:+1:

File: mammon/ext/ircv3/sasl.py (hunk 1 above)
Comment (DanielOaks) [inline mammon/ext/ircv3/sasl.py:62]: Could if we really wanted to, but I don't think users can REG a passphrase
long enough to overrun this right now. (Though, given we accept 2048 rather
than 512... might do it actually. Don't think I took that into account)

Once we have other mechanisms than PLAIN, might be good to allow larger
values too.

**Pull request merged**